runtime/src/accounts.rs

```diff
@@ -267,144 +268,183 @@ impl Accounts {
-                                    {
-                                        account_deps
-                                            .push((programdata_address, programdata_account));
-                                    } else {
-                                        error_counters.account_not_found += 1;
-                                        return Err(TransactionError::ProgramAccountNotFound);
-                                    }
+                                    account_dep_index =
+                                        Some(account_keys.len().saturating_add(account_deps.len())
+                                            as IndexOfAccount);
+                                    account_deps.push((programdata_address, programdata_account));
                                 } else {
-                                    error_counters.invalid_program_for_execution += 1;
-                                    return Err(TransactionError::InvalidProgramForExecution);
+                                    error_counters.account_not_found += 1;
+                                    return Err(TransactionError::ProgramAccountNotFound);
                                 }
+                            } else {
+                                error_counters.invalid_program_for_execution += 1;
+                                return Err(TransactionError::InvalidProgramForExecution);
                             }
-                        } else if account.executable() && message.is_writable(i) {
-                            error_counters.invalid_writable_account += 1;
-                            return Err(TransactionError::InvalidWritableAccount);
                         }
+                    } else if account.executable() && message.is_writable(i) {
+                        error_counters.invalid_writable_account += 1;
+                        return Err(TransactionError::InvalidWritableAccount);
+                    }
 
-                        tx_rent += rent;
-                        rent_debits.insert(key, rent, account.lamports());
+                    tx_rent += rent;
+                    rent_debits.insert(key, rent, account.lamports());
 
-                        account
-                    }
+                    account
                 };
 
+                account_found_and_dep_index.push((account_found, account_dep_index));
                 Ok((*key, account))
             })
             .collect::<Result<Vec<_>>>()?;
 
+        if !validated_fee_payer {
+            error_counters.account_not_found += 1;
+            return Err(TransactionError::AccountNotFound);
+        }
+
         // Appends the account_deps at the end of the accounts,
         // this way they can be accessed in a uniform way.
         // At places where only the accounts are needed,
         // the account_deps are truncated using e.g:
         // accounts.iter().take(message.account_keys.len())
         accounts.append(&mut account_deps);
```

Comment:
Perhaps should remove this line? Looks like `accounts` is expected to not including `account_deps` during iteration of instructions, and Ln 440 appends `account_deps` again.

Reply:
Ah, I should rename the second `account_deps` to `builtins` or something. Yes both are added to `accounts` eventually, but the linear search only runs over the second one which should be smaller because there are only a few built-in programs.